File: src/webapp/src/test/api-client.integration.test.ts
import { createServer, type Server } from 'node:http'
import type { AddressInfo } from 'node:net'
import { QueryClient } from '@tanstack/react-query'
import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest'

import {
  getHealthGetHealthQueryOptions,
  getUsersGetUserQueryOptions,
} from '@/api-client/api'

let apiServer: Server
let queryClient: QueryClient

beforeAll(async () => {
  queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  apiServer = createServer((request, response) => {
    response.setHeader('Content-Type', 'application/json')

    if (request.url === '/api/v1/health') {
      response.end(JSON.stringify({ status: 'ok', version: 'test' }))
      return
    }

    if (
      request.url === '/api/v1/users/user-id' &&
      request.headers.authorization === 'Bearer access-token'
    ) {
      response.end(
        JSON.stringify({
          createdAt: '2026-08-16T00:00:00Z',
          displayName: 'User',
          email: 'user@example.test',
          id: 'user-id',
          updatedAt: '2026-08-16T00:00:00Z',
        }),
      )
      return
    }

    response.statusCode = 401
    response.end(JSON.stringify({ message: 'Unauthorized' }))
  })

  await new Promise<void>((resolve) => {
    apiServer.listen(0, '127.0.0.1', resolve)
  })

  const address = apiServer.address() as AddressInfo
  vi.stubEnv('VITE_API_BASE_URL', `http://127.0.0.1:${address.port}`)
})

afterAll(async () => {
  queryClient.clear()
  vi.unstubAllEnvs()
  await new Promise<void>((resolve, reject) => {
    apiServer.close((error) => {
      if (error) {
        reject(error)
        return
      }

      resolve()
    })
  })
})

describe('generated TanStack Query API integration', () => {
  it('calls the documented public endpoint', async () => {
    const response = await queryClient.fetchQuery(
      getHealthGetHealthQueryOptions(),
    )

    expect(response.data).toEqual({ status: 'ok', version: 'test' })
  })

  it('calls the documented secured endpoint with bearer auth', async () => {
    const response = await queryClient.fetchQuery(
      getUsersGetUserQueryOptions('user-id', {
        fetch: {
          headers: { Authorization: 'Bearer access-token' },
        },
      }),
    )

    expect(response.data.id).toBe('user-id')
  })
})

File: src/mobileapp/test/core/network/api_client_integration_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:api_client/api_client.dart';
import 'package:dio/dio.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mobileapp/core/exceptions/failure.dart';

void main() {
  late HttpServer server;
  late Future<void> Function(HttpRequest request) handleRequest;
  late ApiClient client;

  setUp(() async {
    server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
    server.listen((request) => handleRequest(request));
    client = ApiClient(
      basePathOverride: 'http://${server.address.host}:${server.port}',
    );
  });

  tearDown(() => server.close(force: true));

  test('calls and deserializes the health endpoint', () async {
    handleRequest = (request) async {
      expect(request.method, 'GET');
      expect(request.uri.path, '/api/v1/health');
      request.response.headers.contentType = ContentType.json;
      request.response.write(
        jsonEncode({'status': 'ok', 'version': 'test'}),
      );
      await request.response.close();
    };

    final response = await client.getHealthApi().healthGetHealth();

    expect(response.data?.status, 'ok');
    expect(response.data?.version, 'test');
  });

  test('sends bearer authorization to the user endpoint', () async {
    handleRequest = (request) async {
      expect(request.method, 'GET');
      expect(request.uri.path, '/api/v1/users/user-1');
      expect(
        request.headers.value(HttpHeaders.authorizationHeader),
        'Bearer access-token',
      );
      request.response.headers.contentType = ContentType.json;
      request.response.write(
        jsonEncode({
          'id': 'user-1',
          'email': 'user@example.com',
          'displayName': 'Example User',
          'createdAt': '2026-08-29T00:00:00Z',
          'updatedAt': '2026-08-29T00:00:00Z',
        }),
      );
      await request.response.close();
    };
    client.setApiKey('bearerAuth', 'Bearer access-token');

    final response = await client.getUsersApi().usersGetUser(id: 'user-1');

    expect(response.data?.id, 'user-1');
    expect(response.data?.email, 'user@example.com');
  });

  test(
    'maps generated 401 errors through the shared failure boundary',
    () async {
      handleRequest = (request) async {
        request.response.statusCode = HttpStatus.unauthorized;
        request.response.headers.contentType = ContentType.json;
        request.response.write(jsonEncode({'message': 'Invalid token'}));
        await request.response.close();
      };

      try {
        await client.getUsersApi().usersGetUser(id: 'user-1');
        fail('Expected DioException');
      } on DioException catch (error, stackTrace) {
        final result = Failure.handleException<void>(error, stackTrace);

        result.fold(
          (failure) => expect(failure, isA<UnauthorizedFailure>()),
          (_) => fail('Expected an unauthorized failure'),
        );
      }
    },
  );
}
